### PR TITLE
fix: support unary operator in default value, partition rule and prepare statement

### LINKED ISSUE #4247: Partition rule expr UnaryOp is not a binary expression
### What type of bug is this?

Unexpected error

### What subsystems are affected?

Frontend

### Minimal reproduce step

Create table
```sql
CREATE TABLE `molestiAe`(
`sImiLiQUE` FLOAT NOT NULL,
`amEt` TIMESTAMP(6) TIME INDEX,
`VoLUPtaS` SMALLINT,
`MAGnAm` FLOAT DEFAULT 0.01672709,
tempore DOUBLE,
`Ut` FLOAT NOT NULL,
`EuM` INT NOT NULL,
`sinT` FLOAT NULL,
`dolOReS` BIGINT,
`suNT` SMALLINT,
`aCCuSAntIUM` DOUBLE NOT NULL,
`cOnSequATur` BOOLEAN,
`sEd` DOUBLE NULL,
`qUi` DOUBLE DEFAULT 0.16361014546626762,
`rAtIoNe` FLOAT DEFAULT 0.75340384,
`QuIS` BOOLEAN,
`cupidITAte` FLOAT NOT NULL,
aut BIGINT NULL,
`mOlesTiAS` DOUBLE NULL,
illo DOUBLE NOT NULL,
`proViDENT` SMALLINT,
`eoS` FLOAT DEFAULT 0.2712446,
`EveNIeT` DOUBLE,
quod FLOAT,
`voluPtatibus` INT,
`aSsumenda` INT,
`volUPtAtEm` FLOAT,
et SMALLINT NULL,
`EXpLICaBo` DOUBLE,
PRIMARY KEY(`EveNIeT`, `sImiLiQUE`, `dolOReS`, tempore, `voluPtatibus`, `aSsumenda`, `VoLUPtaS`)
)
PARTITION ON COLUMNS(`sImiLiQUE`) (
`sImiLiQUE` < -113427450000000000000000000000000000000,
`sImiLiQUE` >= -113427450000000000000000000000000000000 AND `sImiLiQUE` < 113427450000000000000000000000000000000,
`sImiLiQUE` >= 113427450000000000000000000000000000000
)
```

### What did you expect to see?

Table created 

### What did you see instead?

```
2024-07-02T07:38:19.486902Z  WARN servers::mysql::server: Internal error occurred during query exec, server actively close the channel to let client try next time err=0: Invalid  │
│ prepare statement: Invalid SQL, error: Partition rule expr UnaryOp { op: Minus, expr: Value(Number("113427450000000000000000000000000000000", false)) } is not a binary expr!, at  │
│ src/servers/src/mysql/handler.rs:487:10 
```

### What operating system did you use?

doesn't matter

### What version of GreptimeDB did you use?

main

### Relevant log output and stack trace

_No response_

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

Closes #4247

## What's changed and what's your intention?

Make the unary operator an optional parameter of `sql_value_to_value` and handle it in several places stated in title.

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
  - Enhanced error handling for unsupported unary operations in SQL statements.
  - Improved control flow in SQL value processing to account for unary operations.
  
- **Bug Fixes**
  - Corrected handling of unary operations to ensure accurate error reporting and processing.

- **Refactor**
  - Streamlined error types to provide clearer feedback on invalid and unsupported unary operations.

### CODE REVIEW

**Review by fengjiachun — APPROVED** [review]

File: src/sql/src/error.rs [hunk 1]
```diff
@@ -285,7 +301,9 @@ impl ErrorExt for Error {
             | ConvertToLogicalExpression { .. }
             | Simplification { .. }
             | InvalidInterval { .. }
-            | PermissionDenied { .. } => StatusCode::InvalidArguments,
+            | PermissionDenied { .. }
```
Comment (killme2008) [inline src/sql/src/error.rs:304]: Not related to this PR, but `PermissionDenied` should return `StatusCode:: PermissionDenied`.

File: src/sql/src/statements.rs [hunk 2]
```diff
@@ -257,6 +259,56 @@ pub fn sql_value_to_value(
...
+            | Value::Float64(_)
+            | Value::Decimal128(_)
+            | Value::Date(_)
+            | Value::DateTime(_)
+            | Value::Timestamp(_)
+            | Value::Time(_)
+            | Value::Duration(_)
+            | Value::Interval(_) => match unary_op {
+                UnaryOperator::Plus => {}
+                UnaryOperator::Minus => value = value.try_negative().unwrap(),
```
Comment (killme2008) [inline src/sql/src/statements.rs:302]: Is the `unwrap` safe here? It seems not.

File: src/sql/src/statements.rs [hunk 3]
```diff
@@ -226,8 +227,9 @@ pub fn sql_value_to_value(
     data_type: &ConcreteDataType,
     sql_val: &SqlValue,
     timezone: Option<&Timezone>,
+    unary_op: Option<UnaryOperator>,
```
Comment (killme2008) [inline src/sql/src/statements.rs:230]: The signature of this function looks ugly, but I don't have a better idea too.

File: src/sql/src/statements.rs [hunk 4]
```diff
@@ -1182,6 +1284,7 @@ mod tests {
             &ConcreteDataType::string_datatype(),
             &SqlValue::Placeholder("default".into()),
             None,
+            None
```
Comment (killme2008) [inline src/sql/src/statements.rs:1287]: Add tests to cover that `unary_op` is not `None`.

**Review by killme2008 — APPROVED** [review]
LGTM

**Comment by killme2008** [comment]
@waynexia After rebasing, there is a compile error that needs to be fixed.
